src/containers/sound-editor.jsx

```diff
@@ -162,52 +236,212 @@ class SoundEditor extends React.Component {
-        this.redoStack.push(this.copyCurrentBuffer());
-        const {samples, sampleRate} = this.undoStack.pop();
+        this.redoStack.push(this.getUndoItem());
+        const {samples, sampleRate, trimStart, trimEnd} = this.undoStack.pop();
         if (samples) {
             this.submitNewSamples(samples, sampleRate, true);
-            this.handlePlay();
+            this.setState({trimStart: trimStart, trimEnd: trimEnd}, this.handlePlay);
         }
     }
     handleRedo () {
-        const {samples, sampleRate} = this.redoStack.pop();
+        const {samples, sampleRate, trimStart, trimEnd} = this.redoStack.pop();
         if (samples) {
-            this.undoStack.push(this.copyCurrentBuffer());
+            this.undoStack.push(this.getUndoItem());
             this.submitNewSamples(samples, sampleRate, true);
-            this.handlePlay();
+            this.setState({trimStart: trimStart, trimEnd: trimEnd}, this.handlePlay);
+        }
+    }
+    handleCopy () {
+        this.copy();
+    }
+    copy (callback) {
+        const trimStart = this.state.trimStart === null ? 0.0 : this.state.trimStart;
+        const trimEnd = this.state.trimEnd === null ? 1.0 : this.state.trimEnd;
+
+        const trimStartSamples = trimStart * this.props.samples.length;
+        const trimEndSamples = trimEnd * this.props.samples.length;
+
+        const newCopyBuffer = this.copyCurrentBuffer();
+        newCopyBuffer.samples = newCopyBuffer.samples.slice(trimStartSamples, trimEndSamples);
+
+        this.setState({
+            copyBuffer: newCopyBuffer
+        }, callback);
+    }
+    handleCopyToNew () {
+        this.copy(this.copyToNew);
+    }
+    copyToNew () {
+        WavEncoder.encode({
+            sampleRate: this.state.copyBuffer.sampleRate,
+            channelData: [this.state.copyBuffer.samples]
+        }).then(wavBuffer => {
+            const vmSound = {
+                format: '',
+                dataFormat: 'wav',
+                rate: this.state.copyBuffer.sampleRate,
+                sampleCount: this.state.copyBuffer.samples.length
+            };
+
+            // Create an asset from the encoded .wav and get resulting md5
+            const storage = this.props.vm.runtime.storage;
+            vmSound.asset = storage.createAsset(
+                storage.AssetType.Sound,
+                storage.DataFormat.WAV,
+                new Uint8Array(wavBuffer),
+                null,
+                true // generate md5
+            );
+            vmSound.assetId = vmSound.asset.assetId;
+
+            // update vmSound object with md5 property
+            vmSound.md5 = `${vmSound.assetId}.${vmSound.dataFormat}`;
+            // The VM will update the sound name to a fresh name
+            vmSound.name = this.props.name;
+
+            this.props.vm.addSound(vmSound);
+        });
+    }
+    resampleBufferToRate (buffer, newRate) {
+        return new Promise(resolve => {
+            if (window.OfflineAudioContext) {
```

Comment:
Maybe we could move this to a static method on AudioEffects so that it can use similar logic with respect to webkitOfflineAudioContext, etc.